x/cron/cron.go

```diff
@@ -0,0 +1,273 @@
+//
+// Due to the implementation details, transaction is guaranteed to be executed
+// after given time, but not exactly at given time. If another transaction is
+// already scheduled for the exact same time, execution of this transaction is
+// delayed until the next free slot.
+//
+// Time granularity is second.
+func (s *Scheduler) Schedule(db weave.KVStore, runAt time.Time, auth []weave.Condition, msg weave.Msg) ([]byte, error) {
+	const granularity = time.Second
+	runAt = runAt.Round(granularity)
+
+	raw, err := s.enc.MarshalTask(auth, msg)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal task")
+	}
+
+	for {
+		key := queueKey(runAt)
+		if ok, err := db.Has(key); err != nil {
+			return nil, errors.Wrap(err, "cannot check key existance")
+		} else if ok {
+			// If the key is already in use, instead of storing a
+			// list of messages under each key, which is a very
+			// unlikely to happen, increase the execution time by
+			// the smallest duration.
+			// Message is guaranteed to be executed not earlier
+			// than given time, NOT at exactly given time.
+			runAt = runAt.Add(granularity)
+			continue
+		}
+
+		if err := db.Set(key, raw); err != nil {
+			return nil, errors.Wrap(err, "cannot store in queue")
+		}
+		return key, nil
+	}
+}
+
+func queueKey(t time.Time) []byte {
+	rawTime := make([]byte, 8)
+	// Zero time does not need to put any data as the bytes are already set
+	// to zero.
+	if !t.IsZero() {
+		binary.BigEndian.PutUint64(rawTime, uint64(t.UnixNano()))
+	}
+	return append([]byte("_crontask:runat:"), rawTime...)
+}
+
+// Delete implements weave.Scheduler interface.
+func (s *Scheduler) Delete(db weave.KVStore, taskID []byte) error {
+	if ok, err := db.Has(taskID); err != nil {
+		return errors.Wrap(err, "has")
+	} else if !ok {
+		return errors.Wrap(errors.ErrNotFound, "no task")
+	}
+	if err := db.Delete(taskID); err != nil {
+		return errors.Wrap(err, "cannot delete")
+	}
+	return nil
+}
+
+// NewTicker returns a cron runner instance that is using given handler to
+// process all queued messages that execution time is due. All serialization is
+// done using provided marshaler.
+//
+// Always use the same marshaler for ticker and scheduler.
+func NewTicker(h weave.Handler, enc TaskMarshaler) *Ticker {
+	return &Ticker{
+		hn:      h,
+		enc:     enc,
+		results: NewTaskResultBucket(),
+	}
+}
+
+// Ticker allows to execute messages queued for future execution. It does this
+// by implementing weave.Ticker interface.
+type Ticker struct {
+	hn      weave.Handler
+	enc     TaskMarshaler
+	results orm.ModelBucket
+}
+
+var _ weave.Ticker = (*Ticker)(nil)
+
+// Tick implementes weave.Ticker interface.
+//
+// Tick can process any number of messages suitable for execution. All changes
+// are done atomically and apply only on success.
+func (t *Ticker) Tick(ctx context.Context, db store.CacheableKVStore) [][]byte {
+	executed, err := t.tick(ctx, db)
+	if err != nil {
```

Comment:
Ahhh... now I see error handling.

If the db write errors (outside of Deliver) cause a panic, no need for a cache.
Writing to the DB will only update the in-memory representation until `Commit` is called (which flushes the whole block to disk).

No need to use cache outside of DeliverTx, as result write errors will lead to a panic, which will not persist state anyway. (I had no idea how to handle failures there either)

Reply:
I understand this is relevant for the `tick` function implementation (https://github.com/iov-one/weave/pull/850/files#diff-fd297d190c42b81d5886e4d0c60bd4f7R173) where I am manually managing database cache and checking every database write.

I can remove the comment part where I mention the database. I think the implementation is correct as it makes sense considered the interfaces it is working with. You know certain scenarios are not possible, because you know the implementation details under the hood :wink: Having that knowledge maybe some shortcuts are possible. I think it is better to keep it aligned with the interfaces instead. This way the code makes more sense to the reader and the functionality is independent of the very specific implementation details of the used components.